File: src/versionable/_migration.py
"""Declarative and imperative migration system.

Migrations transform serialized data from older versions to the current
schema.  They are declared as a ``Migrate`` inner class on Versionable
subclasses::

    @dataclass
    class Config(Versionable, version=3, hash='abc123'):
        name: str
        timeout: int

        class Migrate:
            v2 = Migration().rename('old_name', 'name')
            v1 = Migration().add('timeout', default=30).then(v2)
"""

from __future__ import annotations

import logging
from collections.abc import Callable
from dataclasses import dataclass
from typing import Any

from versionable.errors import MigrationError, UpgradeRequiredError

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Migration operations
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class _RenameOp:
    old: str
    new: str


@dataclass(frozen=True)
class _AddOp:
    field: str
    default: Any


@dataclass(frozen=True)
class _DropOp:
    field: str


@dataclass(frozen=True)
class _ConvertOp:
    field: str
    via: Callable[[Any], Any]
    reverse: Callable[[Any], Any] | None = None


@dataclass(frozen=True)
class _DeriveOp:
    field: str
    fromField: str
    via: Callable[[Any], Any]


@dataclass(frozen=True)
class _SplitOp:
    field: str
    into: dict[str, Callable[[Any], Any]]


@dataclass(frozen=True)
class _MergeOp:
    fields: list[str]
    into: str
    via: Callable[..., Any]


@dataclass(frozen=True)
class _RequiresUpgradeOp:
    pass


type MigrationOp = _RenameOp | _AddOp | _DropOp | _ConvertOp | _DeriveOp | _SplitOp | _MergeOp | _RequiresUpgradeOp


# ---------------------------------------------------------------------------
# Migration class (builder pattern)
# ---------------------------------------------------------------------------


class Migration:
    """Declarative migration builder.

    Chain operations to describe how to transform data from an older
    version to a newer one::

        v2 = (
            Migration()
            .rename('old_name', 'new_name')
            .add('new_field', default=0)
            .drop('removed_field')
        )
    """

    def __init__(self) -> None:
        self._ops: list[MigrationOp] = []

    def rename(self, old: str, new: str) -> Migration:
        """Rename a field."""
        self._ops.append(_RenameOp(old, new))
        return self

    def add(self, field: str, *, default: Any) -> Migration:
        """Add a new field with a default value."""
        self._ops.append(_AddOp(field, default))
        return self

    def drop(self, field: str) -> Migration:
        """Remove a field."""
        self._ops.append(_DropOp(field))
        return self

    def convert(
        self,
        field: str,
        *,
        via: Callable[[Any], Any],
        reverse: Callable[[Any], Any] | None = None,
    ) -> Migration:
        """Convert a field's value using a transformation function."""
        self._ops.append(_ConvertOp(field, via, reverse))
        return self

    def derive(
        self,
        field: str,
        *,
        from_: str,
        via: Callable[[Any], Any],
    ) -> Migration:
        """Derive a new field from an existing one."""
        self._ops.append(_DeriveOp(field, from_, via))
        return self

    def split(
        self,
        field: str,
        *,
        into: dict[str, Callable[[Any], Any]],
    ) -> Migration:
        """Split one field into multiple fields."""
        self._ops.append(_SplitOp(field, into))
        return self

    def merge(
        self,
        fields: list[str],
        *,
        into: str,
        via: Callable[..., Any],
    ) -> Migration:
        """Merge multiple fields into one."""
        self._ops.append(_MergeOp(fields, into, via))
        return self

    def requiresUpgrade(self) -> Migration:
        """Mark this migration as requiring in-place file modification."""
        self._ops.append(_RequiresUpgradeOp())
        return self

    def then(self, other: Migration) -> Migration:
        """Chain another migration after this one."""
        combined = Migration()
        combined._ops = list(self._ops) + list(other._ops)
        return combined

    @property
    def ops(self) -> list[MigrationOp]:
        """Return the list of migration operations (read-only access)."""
        return list(self._ops)


# ---------------------------------------------------------------------------
# Imperative migration decorator
# ---------------------------------------------------------------------------


class _ImperativeMigration:
    """Wrapper for imperative migration functions."""

    def __init__(self, fromVersion: int, fn: Callable[[MigrationContext], None]) -> None:
        self.fromVersion = fromVersion
        self.fn = fn


def migration(fromVersion: int) -> Callable[[Callable[[MigrationContext], None]], _ImperativeMigration]:
    """Decorator for imperative migration functions.

    Usage::

        class Migrate:
            @migration(fromVersion=1)
            def from_v1(ctx: MigrationContext) -> None:
                ctx['new_field'] = ctx.pop('old_field') * 2
    """

    def decorator(fn: Callable[[MigrationContext], None]) -> _ImperativeMigration:
        return _ImperativeMigration(fromVersion, fn)

    return decorator


# ---------------------------------------------------------------------------
# Migration context (for imperative migrations)
# ---------------------------------------------------------------------------


class MigrationContext:
    """Dict-like wrapper for imperative migrations."""

    def __init__(self, data: dict[str, Any]) -> None:
        self._data = data

    def __getitem__(self, key: str) -> Any:
        return self._data[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self._data[key] = value

    def __contains__(self, key: str) -> bool:
        return key in self._data

    def pop(self, key: str, *default: Any) -> Any:
        return self._data.pop(key, *default)

    def drop(self, key: str) -> None:
        self._data.pop(key, None)

    def keys(self) -> Any:
        return self._data.keys()

    def toDict(self) -> dict[str, Any]:
        return dict(self._data)


# ---------------------------------------------------------------------------
# Migration resolution and application
# ---------------------------------------------------------------------------


def resolveMigrations(
    cls: type,
    fromVersion: int,
    toVersion: int,
) -> list[Migration | _ImperativeMigration]:
    """Collect and order migrations from *fromVersion* to *toVersion*.

    Walks the ``Migrate`` inner class on *cls*, collecting ``vN``
    attributes (declarative) and ``@migration``-decorated functions
    (imperative).  Migrations are ordered from oldest to newest.
    """
    migrateClass = getattr(cls, "Migrate", None)
    if migrateClass is None:
        raise MigrationError(
            f"{cls.__qualname__} version {toVersion} requires migration from "
            f"v{fromVersion}, but no Migrate class is defined."
        )

    # Collect all migrations with their version numbers
    migrations: dict[int, Migration | _ImperativeMigration] = {}

    for name in dir(migrateClass):
        attr = getattr(migrateClass, name)

        # Declarative: vN attributes
        if name.startswith("v") and name[1:].isdigit() and isinstance(attr, Migration):
            version = int(name[1:])
            migrations[version] = attr

        # Imperative: @migration decorated functions
        if isinstance(attr, _ImperativeMigration):
            migrations[attr.fromVersion] = attr

    # Build ordered chain: fromVersion → fromVersion+1 → ... → toVersion-1
    chain: list[Migration | _ImperativeMigration] = []
    for v in range(fromVersion, toVersion):
        if v not in migrations:
            raise MigrationError(
                f"No migration from v{v} to v{v + 1} defined on {cls.__qualname__}.Migrate (expected 'v{v}' attribute)."
            )
        chain.append(migrations[v])

    return chain


def applyMigrations(
    data: dict[str, Any],
    migrations: list[Migration | _ImperativeMigration],
    *,
    upgradeInPlace: bool = False,
) -> dict[str, Any]:
    """Apply a chain of migrations to *data*.

    Returns the migrated data dict.
    """
    result = dict(data)

    for mig in migrations:
        if isinstance(mig, _ImperativeMigration):
            ctx = MigrationContext(result)
            mig.fn(ctx)
            result = ctx.toDict()
        elif isinstance(mig, Migration):
            result = _applyDeclarativeMigration(result, mig, upgradeInPlace=upgradeInPlace)
        else:
            raise MigrationError(f"Unknown migration type: {type(mig)}")

    return result


def _applyDeclarativeMigration(
    data: dict[str, Any],
    mig: Migration,
    *,
    upgradeInPlace: bool = False,
) -> dict[str, Any]:
    """Apply a single declarative migration's operations."""
    result = dict(data)

    for op in mig.ops:
        if isinstance(op, _RenameOp):
            if op.old in result:
                result[op.new] = result.pop(op.old)

        elif isinstance(op, _AddOp):
            if op.field not in result:
                default = op.default
                if callable(default):
                    default = default()
                result[op.field] = default

        elif isinstance(op, _DropOp):
            result.pop(op.field, None)

        elif isinstance(op, _ConvertOp):
            if op.field in result:
                result[op.field] = op.via(result[op.field])

        elif isinstance(op, _DeriveOp):
            if op.fromField in result:
                result[op.field] = op.via(result[op.fromField])

        elif isinstance(op, _SplitOp):
            if op.field in result:
                source = result.pop(op.field)
                for targetField, fn in op.into.items():
                    result[targetField] = fn(source)

        elif isinstance(op, _MergeOp):
            values = [result.pop(f) for f in op.fields if f in result]
            if values:
                result[op.into] = op.via(*values)

        elif isinstance(op, _RequiresUpgradeOp) and not upgradeInPlace:
            raise UpgradeRequiredError(
                "This migration requires in-place file modification. "
                "Use upgradeInPlace=True or call upgradeFile() first."
            )

    return result

File: dotnet/src/Versionable/Migrations/Migration.cs
using Versionable.Errors;

namespace Versionable.Migrations;

/// <summary>
/// A declarative migration: an ordered list of <see cref="MigrationOp"/> that reshapes the raw
/// field dictionary of one schema version into the next.
/// </summary>
/// <remarks>
/// Python counterpart: the <c>Migration</c> builder in <c>src/versionable/_migration.py</c>, op
/// for op. Declared as a member of the nested <c>Migrate</c> class named for the version it reads:
/// <code>
/// public static class Migrate
/// {
///     public static readonly Migration V1 = new Migration().Rename("title", "name").Drop("debug");
///     public static readonly Migration V2 = new Migration().Add("timeout_ms", 0);
/// }
/// </code>
/// <para>
/// <b>Each builder call returns a new instance</b> rather than mutating the receiver, which
/// Python's does. The op semantics are identical; what differs is that a <c>Migration</c> published
/// as a <c>static readonly</c> field — the way every declared migration is published — cannot be
/// extended by one caller into something a later caller sees. A partly built migration that is
/// never assigned is simply dropped, whereas Python's would silently keep the ops.
/// </para>
/// </remarks>
public sealed class Migration
{
    private static readonly MigrationOp[] _empty = [];

    private readonly MigrationOp[] _ops;

    /// <summary>Initializes a new instance of the <see cref="Migration"/> class with no operations.</summary>
    public Migration()
        : this(_empty)
    {
    }

    private Migration(MigrationOp[] ops)
    {
        _ops = ops;
        Ops = Array.AsReadOnly(ops);
    }

    /// <summary>The operations, in the order they were declared and are applied.</summary>
    /// <remarks>
    /// A read-only wrapper rather than the backing array, which a caller could otherwise cast
    /// back to <c>MigrationOp[]</c> and rewrite. Python hands back a copy for the same reason.
    /// </remarks>
    public IReadOnlyList<MigrationOp> Ops { get; }

    /// <summary>Moves a value from one wire name to another.</summary>
    /// <param name="oldName">Wire name the old file used.</param>
    /// <param name="newName">Wire name the current schema uses.</param>
    /// <returns>A migration with the operation appended.</returns>
    public Migration Rename(string oldName, string newName)
    {
        ArgumentNullException.ThrowIfNull(oldName);
        ArgumentNullException.ThrowIfNull(newName);
        return With(new RenameOp(oldName, newName));
    }

    /// <summary>Adds a field that files at the older version do not carry.</summary>
    /// <param name="field">Wire name to add.</param>
    /// <param name="value">Value to give it. A field already present keeps its own value.</param>
    /// <returns>A migration with the operation appended.</returns>
    /// <remarks>
    /// The value is stored as-is, delegate or not: a delegate passed here becomes the field's
    /// value rather than being called, which no backend can write. Use
    /// <see cref="AddComputed(string, Func{object})"/> for a value computed per application —
    /// Python's <c>add(field, default=callable)</c>.
    /// </remarks>
    public Migration Add(string field, object? value)
    {
        ArgumentNullException.ThrowIfNull(field);
        return With(new AddOp(field, value, null));
    }

    /// <summary>Adds a field whose value is produced per application.</summary>
    /// <param name="field">Wire name to add.</param>
    /// <param name="defaultFactory">Produces the value each time the migration runs.</param>
    /// <returns>A migration with the operation appended.</returns>
    /// <remarks>
    /// Python counterpart: <c>add(field, default=list)</c> — a callable default, which Python calls
    /// rather than stores. Spelled as its own method because C# 10 gave lambdas a natural
    /// conversion to <see cref="object"/>, so an overload of <see cref="Add(string, object?)"/>
    /// would be ambiguous at every call site.
    /// </remarks>
    public Migration AddComputed(string field, Func<object?> defaultFactory)
    {
        ArgumentNullException.ThrowIfNull(field);
        ArgumentNullException.ThrowIfNull(defaultFactory);
        return With(new AddOp(field, null, defaultFactory));
    }

    /// <summary>Removes a field the current schema no longer has.</summary>
    /// <param name="field">Wire name to remove. Absence is not an error.</param>
    /// <returns>A migration with the operation appended.</returns>
    public Migration Drop(string field)
    {
        ArgumentNullException.ThrowIfNull(field);
        return With(new DropOp(field));
    }

    /// <summary>Rewrites a field's value in place.</summary>
    /// <param name="field">Wire name to convert. Absence is not an error.</param>
    /// <param name="via">Maps the old value to the new one.</param>
    /// <param name="reverse">The inverse, recorded for a future downgrade path and never run.</param>
    /// <returns>A migration with the operation appended.</returns>
    public Migration Convert(
        string field,
        Func<object?, object?> via,
        Func<object?, object?>? reverse = null)
    {
        ArgumentNullException.ThrowIfNull(field);
        ArgumentNullException.ThrowIfNull(via);
        return With(new ConvertOp(field, via, reverse));
    }

    /// <summary>Computes a new field from an existing one, which stays in place.</summary>
    /// <param name="field">Wire name to write.</param>
    /// <param name="fromField">Wire name to read. When absent, nothing is written.</param>
    /// <param name="via">Maps the source value to the derived one.</param>
    /// <returns>A migration with the operation appended.</returns>
    public Migration Derive(string field, string fromField, Func<object?, object?> via)
    {
        ArgumentNullException.ThrowIfNull(field);
        ArgumentNullException.ThrowIfNull(fromField);
        ArgumentNullException.ThrowIfNull(via);
        return With(new DeriveOp(field, fromField, via));
    }

    /// <summary>Replaces one field with several derived from it.</summary>
    /// <param name="field">Wire name to consume. When absent, nothing happens.</param>
    /// <param name="targets">Fields to write, in order.</param>
    /// <returns>A migration with the operation appended.</returns>
    /// <remarks>
    /// Python passes a dict, whose insertion order it preserves; the targets are an ordered list
    /// here because a C# dictionary's enumeration order is not part of its contract, and the phase-6
    /// converter has to emit the same order it read.
    /// </remarks>
    public Migration Split(string field, params SplitTarget[] targets)
    {
        ArgumentNullException.ThrowIfNull(field);
        ArgumentNullException.ThrowIfNull(targets);
        return With(new SplitOp(field, (SplitTarget[])targets.Clone()));
    }

    /// <summary>Replaces several fields with one combined from them.</summary>
    /// <param name="fields">Wire names to consume, in the order <paramref name="via"/> expects.</param>
    /// <param name="into">Wire name to write.</param>
    /// <param name="via">Combines the values of the fields that were present.</param>
    /// <returns>A migration with the operation appended.</returns>
    public Migration Merge(IReadOnlyList<string> fields, string into, Func<object?[], object?> via)
    {
        ArgumentNullException.ThrowIfNull(fields);
        ArgumentNullException.ThrowIfNull(into);
        ArgumentNullException.ThrowIfNull(via);
        return With(new MergeOp([.. fields], into, via));
    }

    /// <summary>Declares that this migration rewrites the file rather than only the loaded data.</summary>
    /// <returns>A migration with the operation appended.</returns>
    public Migration RequiresUpgrade() => With(new RequiresUpgradeOp());

    /// <summary>Concatenates another migration's operations after this one's.</summary>
    /// <param name="other">The migration to run second.</param>
    /// <returns>A migration holding both op lists, leaving both operands unchanged.</returns>
    public Migration Then(Migration other)
    {
        ArgumentNullException.ThrowIfNull(other);

        MigrationOp[] combined = new MigrationOp[_ops.Length + other._ops.Length];
        Array.Copy(_ops, combined, _ops.Length);
        Array.Copy(other._ops, 0, combined, _ops.Length, other._ops.Length);
        return new Migration(combined);
    }

    /// <summary>Runs every operation over <paramref name="fields"/>, in declaration order.</summary>
    /// <param name="fields">
    /// Raw field values keyed by wire name, envelope keys already stripped. Mutated in place and
    /// returned: unlike Python's <c>applyMigrations</c>, which copies the dict once per step, the
    /// dictionary handed to a chain is built by the load path for that one load and has no other
    /// reader.
    /// </param>
    /// <param name="upgradeInPlace">Whether the caller has permitted file-rewriting migrations.</param>
    /// <returns><paramref name="fields"/>, reshaped.</returns>
    /// <exception cref="UpgradeRequiredException">
    /// The migration declares <see cref="RequiresUpgrade"/> and <paramref name="upgradeInPlace"/>
    /// is <see langword="false"/>.
    /// </exception>
    public IDictionary<string, object?> Apply(IDictionary<string, object?> fields, bool upgradeInPlace = false)
    {
        ArgumentNullException.ThrowIfNull(fields);

        foreach (MigrationOp op in _ops)
        {
            ApplyOp(op, fields, upgradeInPlace);
        }

        return fields;
    }

    private static void ApplyOp(MigrationOp op, IDictionary<string, object?> fields, bool upgradeInPlace)
    {
        switch (op)
        {
            case RenameOp rename:
                if (fields.TryGetValue(rename.OldName, out object? renamed))
                {
                    fields.Remove(rename.OldName);
                    fields[rename.NewName] = renamed;
                }

                break;

            case AddOp add:
                if (!fields.ContainsKey(add.Field))
                {
                    fields[add.Field] = add.Factory is null ? add.Value : add.Factory();
                }

                break;

            case DropOp drop:
                fields.Remove(drop.Field);
                break;

            case ConvertOp convert:
                if (fields.TryGetValue(convert.Field, out object? converting))
                {
                    fields[convert.Field] = convert.Via(converting);
                }

                break;

            case DeriveOp derive:
                if (fields.TryGetValue(derive.FromField, out object? source))
                {
                    fields[derive.Field] = derive.Via(source);
                }

                break;

            case SplitOp split:
                if (fields.TryGetValue(split.Field, out object? splitting))
                {
                    fields.Remove(split.Field);
                    foreach (SplitTarget target in split.Targets)
                    {
                        fields[target.Field] = target.Via(splitting);
                    }
                }

                break;

            case MergeOp merge:
                ApplyMerge(merge, fields);
                break;

            case RequiresUpgradeOp when !upgradeInPlace:
                throw new UpgradeRequiredException(
                    "This migration requires in-place file modification. Load with "
                        + "VersionableLoadOptions { UpgradeInPlace = true } to permit it.");

            case RequiresUpgradeOp:
                break;

            default:
                throw new MigrationException($"Unknown migration operation: {op.GetType().Name}.");
        }
    }

    private static void ApplyMerge(MergeOp merge, IDictionary<string, object?> fields)
    {
        // Python collects only the fields that are present and passes them positionally, so a file
        // missing one of them merges the rest rather than failing. Same here, short array and all.
        List<object?> present = new(merge.Fields.Count);
        foreach (string field in merge.Fields)
        {
            if (fields.TryGetValue(field, out object? value))
            {
                fields.Remove(field);
                present.Add(value);
            }
        }

        if (present.Count > 0)
        {
            fields[merge.Into] = merge.Via([.. present]);
        }
    }

    private Migration With(MigrationOp op)
    {
        MigrationOp[] extended = new MigrationOp[_ops.Length + 1];
        Array.Copy(_ops, extended, _ops.Length);
        extended[_ops.Length] = op;
        return new Migration(extended);
    }
}
